main/report.py

```diff
@@ -0,0 +1,280 @@
+        start_date: the start date for the charging period
+        end_date: the end date for the charging period
+
+    Returns:
+        A tuple of the number of chargeable days and list of pks of relevant time
+            entries, or a tuple of None values if there are no time entries.
+    """
+    start_time = datetime.combine(start_date, datetime.min.time())
+    end_time = datetime.combine(end_date, datetime.min.time())
+
+    time_entries = models.TimeEntry.objects.filter(
+        project=project,
+        start_time__gte=start_time,
+        start_time__lt=end_time,
+    )
+    pks = list(time_entries.values_list("pk", flat=True))
+
+    if len(time_entries) == 0:
+        return None, None
+
+    hours, _ = utils.get_logged_hours(time_entries)
+    total_days = round(hours / 7, 3)
+    return total_days, pks
+
+
+def create_monthly_charges(project: Project, start_date: date, end_date: date) -> None:
+    """Create monthly charges for projects with Actual and Pro-rata charging.
+
+    Args:
+        project: the Project for charging
+        start_date: the start date for the charging period
+        end_date: the end date for the charging period
+    """
+    # get the amount of chargeable days for Actual or Pro-rata projects
+    if project.charging == "Actual":  # get pk values for the relevant time entries
+        total_days, pks = get_actual_chargeable_days(project, start_date, end_date)
+    if project.charging == "Pro-rata":
+        total_days = get_pro_rata_chargeable_days(project, start_date, end_date)
+        pks = None
+
+    if total_days and project.days_left:
+        if total_days > project.days_left[0]:
+            raise ValidationError(
+                "Total chargeable days exceeds the total effort left "
+                f"for project {project.name}."
+            )
+
+        # get non-expired funding with funding left
+        funding_sources = list(
+            project.funding_source.all()
+            .filter(
+                expiry_date__gte=end_date,
+            )
+            .order_by("expiry_date")
+        )
+        funding_sources = [
+            funding for funding in funding_sources if funding.funding_left > 0
+        ]
+
+        # create a monthly charge for each funding source
+        for funding in funding_sources:
+            if total_days > 0:  # if there are days left to charge
+                days_deduce = min(total_days, funding.effort_left)
+                amount = round(days_deduce * float(funding.daily_rate), 2)
+                charge = models.MonthlyCharge.objects.create(
+                    project=project, funding=funding, amount=amount, date=start_date
+                )
+                total_days -= days_deduce
+
+                # update time entries with monthly charge
+                if pks:
+                    for time_entry in models.TimeEntry.objects.filter(pk__in=pks):
+                        time_entry.monthly_charge.add(charge)
```

Comment:
An stylistic suggestion: when you have nesting it is often more readable to reverse the logic and unindent the whole block:

```suggestion
        for funding in funding_sources:
            if total_days <= 0:  # We are done charging
                break

            days_deduce = min(total_days, funding.effort_left)
            amount = round(days_deduce * float(funding.daily_rate), 2)
            charge = models.MonthlyCharge.objects.create(
                project=project, funding=funding, amount=amount, date=start_date
            )
            total_days -= days_deduce

            # update time entries with monthly charge
            if pks:
                for time_entry in models.TimeEntry.objects.filter(pk__in=pks):
                    time_entry.monthly_charge.add(charge)
```